doc/flow-definition.md

```diff
@@ -1,11 +1,308 @@
+> * It's confusing that blocks of code are not necessarily executed in the order that they appear in the flow definition.
+> * The code in the blocks is not easily unit-testable.
+
+Each flow is instantiated in the `FlowRegistry` via the `Flow.build` method which, in turn, instantiates the flow and invokes its `Flow#define` method.
+
+Currently in production a single instance of each flow class is instantiated at application start-up (or strictly speaking on the first request) and cached in the `FlowRegistry` i.e. the same instance of a particular flow class is used to service all the user requests involving that Smart Answer. Thus it's important that no request-specific state is stored on the flow instance, otherwise this could leak into other user requests.
+
+It's important to understand this distinction between "flow definition time" and "request time" when trying to understand a flow definition. A number of the sections below refer to this distinction.
+
+### Metadata
+
+There are a number of methods on `SmartAnswer::Flow` which allow "metadata" for the flow to be specified:
+
+```ruby
+module SmartAnswer
+  class ExampleSmartAnswerFlow < Flow
+    def define
+      name 'example-smart-answer' # this is the path where the Smart Answer will be registered on gov.uk (via Panopticon)
+      content_id "bfda3b4f-166b-48e7-9aaf-21bfbd606207" # a UUID used by v2 of the Publishing API (?)
+      status :published # this indicates whether or not the flow should appear on gov.uk (i.e. production); those with `:draft` status will not appear
+      satisfies_need "123456" # may relate the Smart Answer to the original user need in the Need-o-tron app (?)
+
+      # question & outcome definitions specified here
+    end
+  end
+end
+```
+
+### Arbitrary Ruby code
+
+Since the flow definition is just the body of a standard Ruby method, it's possible to write arbitrary Ruby code at any point within it.
+
+> It should be possible to implement the majority of flows *without* writing such arbitrary Ruby code. Any decision to introduce such arbitrary code should be very carefully considered, particularly regarding it's maintainability.
+
+Since the flow definition is just the body of a `Flow` instance method (`#define`), the value of `self` at the "top-level" within the method is an instance of the flow.
+
+#### Local variables
+
+Some flows (e.g. [register-a-birth][local-variable-in-flow-definition]) define local variables towards the top of the flow definition. Standard Ruby scoping means that these are then available within question & outcome blocks, as well as within blocks nested inside these blocks. However, note that the value of the local variable is set at flow definition time and *not* at request time. Since a single flow instance is cached in the `FlowRegistry` at application start-up time, this means the local variable is _only assigned once_.
+
```

Comment:
Yes. I've tried to made this clearer in this fix-up commit: 1c4acfa1f6b5b439361e0c1fc7b943326ea53618.